src/Swarm/Game/Step.hs

```diff
@@ -752,6 +753,7 @@ execConst c vs s k = do
         Just n -> fromMaybe e <$> uses entityMap (lookupEntityName n)
 
       robotInventory %= insert e'
+      updateDiscoveredEntities e'
```

Comment:
I worry that in the future when anyone writes code to adjust the `robotInventory` or installed devices, they will forget to call the proper `update` function, or not even realize that they are supposed to.  (And indeed, how can we be confident that you have found all the right places to insert calls to these functions?)  Could we somehow refactor into *e.g.* a single function that both updates the robot inventory and also updates the discovered entity set appropriately?

Reply:
I wanted to keep the overhead of this nice-to-have feature as low as possible, and I figured we only need to update the lists on Make, Create and Grab, any other inventory update could go without updating the new lists (because the entity would already be known). On the other hand, having a common function might be useful for calculating achievements in the future, but unless the overhead is minimal, I think this should be only used by Make, Create and Grab.

Reply:
But what about `give`?  i.e. suppose you program a robot to go grab a thing that you don't know about and then give it to you.  Shouldn't that be able to trigger the availability of a new recipe?  What about if you `salvage` that robot instead of having it `give` you something? There are many, many ways a robot's inventory can gain new entities (I'm quite sure I'm forgetting some).

Reply:
I initially implemented something along those lines, but it sounded like a lot of work (to check the current or receiving robot inventory). Instead the current solution only keeps track of a single global list of known entities, so that the available recipes list is still updated, even if the current robot does not have all the ingredients. The reason being that if the player acquired an entity by any mean, then the player would know how to gather the ingredients for making the new recipe.

Reply:
Ah, I see now, that makes sense.

Reply:
Though I agree that the present solution is not great either. But perhaps if we miss a case, then that's not the end of the world, the player may still go through the inventory to check the full list.